### ISSUE TITLE
function cannot be called in log phase (only in: rewrite, access, response, header_filter)

### ISSUE THREAD

**Issue by nvx** [Kong/kong#7065]
### Summary

Getting the following intermittently in the Kong error log after upgrading to 2.4.0 (from 2.1.4):

```
2021/04/20 16:34:11 [error] 2356#0: *4061 lua entry thread aborted: runtime error: /usr/local/share/lua/5.1/kong/pdk/private/phases.lua:90: function cannot be called in log phase (only in: rewrite, access, response, header_filter)
stack traceback:
coroutine 0:
        [C]: in function 'error'
        /usr/local/share/lua/5.1/kong/pdk/private/phases.lua:90: in function 'check_phase'
        /usr/local/share/lua/5.1/kong/pdk/response.lua:794: in function 'exit'
        /usr/local/share/lua/5.1/kong/init.lua:319: in function 'access'
        access_by_lua(nginx-kong.conf:95):2: in main chunk, client: x.x.x.x, server: kong, request: "GET /some/api HTTP/2.0", host: "api.example.com", referrer: "https://example.com/"
```

Responses to HTTP requests have a HTTP 500 with the following body:

```
{
  "message":"An unexpected error occurred"
}
```

Most requests seem fine (even to the same API) so I'm not sure what triggers it to occur for some requests and not others.

Due to the overly generic nature of the error message I'm not sure where to start looking for additional information.

### Additional Details & Logs

- Kong version (`$ kong version`)
- Kong debug-level startup logs 2.4.0
- Kong error logs:
```
2021/04/20 16:28:41 [notice] 1843#0: using the "epoll" event method
2021/04/20 16:28:41 [notice] 1843#0: openresty/1.19.3.1
2021/04/20 16:28:41 [notice] 1843#0: built by gcc 4.8.5 20150623 (Red Hat 4.8.5-44) (GCC)
2021/04/20 16:28:41 [notice] 1843#0: OS: Linux 3.10.0-1160.21.1.el7.x86_64
2021/04/20 16:28:41 [notice] 1843#0: getrlimit(RLIMIT_NOFILE): 1048576:1048576
2021/04/20 16:28:41 [notice] 1843#0: start worker processes
2021/04/20 16:28:41 [notice] 1843#0: start worker process 2355
2021/04/20 16:28:41 [notice] 1843#0: start worker process 2356
2021/04/20 16:28:41 [notice] 2355#0: *1 [lua] init.lua:260: purge(): [DB cache] purging (local) cache, context: init_worker_by_lua*
2021/04/20 16:28:41 [notice] 2355#0: *1 [lua] init.lua:260: purge(): [DB cache] purging (local) cache, context: init_worker_by_lua*
2021/04/20 16:28:41 [notice] 2356#0: *2 [lua] globalpatches.lua:55: sleep(): executing a blocking 'sleep' (0.001 seconds), context: init_worker_by_lua*
2021/04/20 16:28:41 [notice] 2356#0: *2 [lua] globalpatches.lua:55: sleep(): executing a blocking 'sleep' (0.002 seconds), context: init_worker_by_lua*
2021/04/20 16:28:41 [notice] 2356#0: *2 [lua] globalpatches.lua:55: sleep(): executing a blocking 'sleep' (0.004 seconds), context: init_worker_by_lua*
2021/04/20 16:28:41 [notice] 2356#0: *2 [lua] globalpatches.lua:55: sleep(): executing a blocking 'sleep' (0.008 seconds), context: init_worker_by_lua*
2021/04/20 16:28:41 [notice] 2356#0: *2 [lua] globalpatches.lua:55: sleep(): executing a blocking 'sleep' (0.016 seconds), context: init_worker_by_lua*
2021/04/20 16:28:41 [notice] 2356#0: *2 [lua] globalpatches.lua:55: sleep(): executing a blocking 'sleep' (0.032 seconds), context: init_worker_by_lua*
2021/04/20 16:28:41 [notice] 2356#0: *2 [lua] globalpatches.lua:55: sleep(): executing a blocking 'sleep' (0.064 seconds), context: init_worker_by_lua*
2021/04/20 16:28:41 [notice] 2355#0: *1 [kong] init.lua:371 declarative config loaded from /etc/kong/kong.yaml, context: init_worker_by_lua*
2021/04/20 16:28:42 [notice] 2356#0: *2 [kong] init.lua:214 only worker #0 can manage, context: init_worker_by_lua*
```
- Kong configuration (the output of a GET request to Kong's Admin port - see
  https://docs.konghq.com/latest/admin-api/#retrieve-node-information)
- Operating system
RedHat 7 x64

**Comment by nvx** [comment]
I noticed it also occurs with a slightly different stack trace too:

```
2021/04/20 15:04:28 [error] 30091#0: *10116106 lua entry thread aborted: runtime error: /usr/local/share/lua/5.1/kong/pdk/private/phases.lua:90: function cannot be called in log phase (only in: rewrite, access, response, header_filter)
stack traceback:
coroutine 0:
        [C]: in function 'error'
        /usr/local/share/lua/5.1/kong/pdk/private/phases.lua:90: in function 'check_phase'
        /usr/local/share/lua/5.1/kong/pdk/response.lua:794: in function 'after'
        /usr/local/share/lua/5.1/kong/init.lua:780: in function 'access'
        access_by_lua(nginx-kong.conf:95):2: in main chunk
```

I'm rather confused as it seems to be being called from within the access_by_lua directive, which says to me that the phase is access, so the error is rather confusing saying it cannot be called in the log phase?

**Comment by nvx** [comment]
Reverting to 2.3.3 seems to have resolved the issue, so looks like a regression in 2.4.0 specifically not any of the other interim versions between 2.1.4 which was running previously.

**Comment by bungle** [comment]
Do you have any plugins that implement `response` phase?

**Comment by bungle** [comment]
This looks wrong though:
```
https://github.com/Kong/kong/blob/master/kong/pdk/response.lua#L794
```

It should include `response` phase. But the error is interesting too `function cannot be called in log phase`.

**Comment by bungle** [comment]
Also do you use Websockets?

**Comment by nvx** [comment]
> This looks wrong though:
> https://github.com/Kong/kong/blob/master/kong/pdk/response.lua#L794
> 
> It should include `response` phase. But the error is interesting too `function cannot be called in log phase`.

The error does specifically mention it includes the response phase at the very end `(only in: rewrite, access, response, header_filter)` (gotta scroll across in GitHub) so I think that line is right, but somehow it's getting confused about what phase it's in thinking it's log when it should be access (the very top of the call chain is Kong.access, which is called by access_by_lua so my understanding is the phase should be access not response at that point? Both access and response are well before log though).

The definition of that `rewrite_access_header` is here https://github.com/Kong/kong/blob/911385466571e4f72999fa43419daafae08159d0/kong/pdk/response.lua#L51-L56 and includes response.

> Do you have any plugins that implement `response` phase or calls `kong.service.request.enable_buffering()`?

I use the following plugins:
* response-transformer
* request-transformer
* request-termination (only to reject some requests, wouldn't have been triggered in the requests where this issue was encountered)
* udp-log
* prometheus
* correlation-id
* key-auth
* basic-auth

I imagine the response-transformer plugin would use the response phase (I use the add and remove headers directives to add stuff like HSTS headers to every response), perhaps correlation-id too (not sure if this adds it to the response too off hand or only the request).

Not sure about enable_buffering calls, looking at the error.log some requests do result in buffering due to large request or more often large response sizes. The requests causing this issue were quite small though (GET requests, no request body data) so shouldn't have triggered buffering, and looking at the backend API logs I don't think the request made it to the backend, so there would have been no response from the backend to trigger buffering to disk either though (at any rate those specific calls weren't expected to return bulk data anyway, only some tiny JSON bodies).

> Also do you use Websockets?

Not at all. I do use HTTP/2 though as most of my stack is pretty modern (clients are either Golang or web browsers, both of which happily do HTTP/2).

**Comment by kikito** [comment]
Hello @nvx, while I was not able to reproduce your exact issue locally, I have detected some problems with the phase checker (the code that usually provides a warning when a function is used out-of-place in the PDK, providing a safe runtime error) which are quite probably related with your issue in #7109.

The plan is merging the fix into the next bugfix version of Kong, which should be coming soon. Thank you for your patience.

**Comment by nvx** [comment]
It seemed to occur fairly frequently for me even in test environments, so will be pretty easy to check if that fixes it once the new version is cut and go from there.

**Comment by nvx** [comment]
I notice 2.4.1 is out now. I've updated a couple of test hosts, will monitor it over the next few days to see if the issue reoccurs, although so far a few hours in I haven't seen any occurrences while on 2.4.0 I saw it within minutes so it's looking positive that #7109 has fixed the issue so far.

**Comment by nvx** [comment]
Still no reoccurrence, so I'll close this off as it seems to be fixed in 2.4.1. Cheers @kikito!
